### PR TITLE
build: don't minimize CRD for `WorkflowEventBinding` and `WorkflowArtifactGCTask`. Fixes #12166

### LINKED ISSUE #12166: `WorkflowEventBinding` CRD has no documentation
# Summary

It is customary for K8S CRDs to be self documenting, enabling one to run `kubectl explain <CRD path>` (or use your IDE's features to do so) to assist in defining K8S resources. This is missing from `WorkflowEventBinding`, making it unnecessarily burdensome to figure out proper usage.

Here's the current output when running `kubectl explain WorkflowEventBinding.spec`.

```text
GROUP:      argoproj.io
KIND:       WorkflowEventBinding
VERSION:    v1alpha1

FIELD: spec <Object>

DESCRIPTION:
    <empty>

```

Note: This may apply to other Argo CRDs; this is just the one I'm attempting to use right now.

## Use Cases

When implementing Argo event bindings.

---

**Message from the maintainers**:

Love this enhancement proposal? Give it a 👍. We prioritise the proposals with the most 👍.

### PR DESCRIPTION
**Pull request by MasonM**
Fixes #12166. Partial fix for #8190

### Motivation

The CRDs under `manifests/base/crds/minimal/` are minimized as a workaround for https://github.com/kubernetes/kubernetes/issues/82292. However, `WorkflowEventBinding` and `WorkflowArtifactGCTask` are small enough that minimization isn't necessary. Including the full CRD enables users to use `kubectl explain` to explore the CRD definitions.

### Modifications

* Rename `removeCRDValidation()` to `minimizeCRD()` so it more accurately describes what it's doing and add comments
* Update `minimizeCRD()` to only minimize `cronworkflows.argoproj.io`, `clusterworkflowtemplates.argoproj.io`, `workflows.argoproj.io`, `workflowtemplates.argoproj.io`, `workflowtasksets.argoproj.io`, since the others are small enough they don't need to be minimized
* Add more details to `manifests/base/crds/full/README.md` and `manifests/base/crds/minimal/README.md`

### Verification
Ran `make start` and verified `kubectl explain` returns spec details for the CRDs:
<details>
<summary>Click here for <code>kubectl explain --recursive WorkflowEventBinding</code> output</summary>

```
$ kubectl explain --recursive WorkflowEventBinding
GROUP:      argoproj.io
KIND:       WorkflowEventBinding
VERSION:    v1alpha1

DESCRIPTION:
    <empty>
FIELDS:
  apiVersion    <string>
  kind  <string>
  metadata      <ObjectMeta> -required-
    annotations <map[string]string>
    creationTimestamp   <string>
    deletionGracePeriodSeconds  <integer>
    deletionTimestamp   <string>
    finalizers  <[]string>
    generateName        <string>
    generation  <integer>
    labels      <map[string]string>
    managedFields       <[]ManagedFieldsEntry>
      apiVersion        <string>
      fieldsType        <string>
      fieldsV1  <FieldsV1>
      manager   <string>
      operation <string>
      subresource       <string>
      time      <string>
    name        <string>
    namespace   <string>
    ownerReferences     <[]OwnerReference>
      apiVersion        <string> -required-
      blockOwnerDeletion        <boolean>
      controller        <boolean>
      kind      <string> -required-
      name      <string> -required-
      uid       <string> -required-
    resourceVersion     <string>
    selfLink    <string>
    uid <string>
  spec  <Object> -required-
    event       <Object> -required-
      selector  <string> -required-
    submit      <Object>
      arguments <Object>
        artifacts       <[]Object>
          archive       <Object>
            none        <Object>
            tar <Object>
              compressionLevel  <integer>
            zip <Object>
          archiveLogs   <boolean>
          artifactGC    <Object>
            podMetadata <Object>
              annotations       <map[string]string>
              labels    <map[string]string>
            serviceAccountName  <string>
            strategy    <string>
          artifactory   <Object>
            passwordSecret      <Object>
              key       <string> -required-
              name      <string>
              optional  <boolean>
            url <string> -required-
            usernameSecret      <Object>
              key       <string> -required-
              name      <string>
              optional  <boolean>
          azure <Object>
            accountKeySecret    <Object>
              key       <string> -required-
              name      <string>
              optional  <boolean>
            blob        <string> -required-
            container   <string> -required-
            endpoint    <string> -required-
            useSDKCreds <boolean>
          deleted       <boolean>
          from  <string>
          fromExpression        <string>
          gcs   <Object>
            bucket      <string>
            key <string> -required-
            serviceAccountKeySecret     <Object>
              key       <string> -required-
              name      <string>
              optional  <boolean>
          git   <Object>
            branch      <string>
            depth       <integer>
            disableSubmodules   <boolean>
            fetch       <[]string>
            insecureIgnoreHostKey       <boolean>
            passwordSecret      <Object>
              key       <string> -required-
              name      <string>
              optional  <boolean>
            repo        <string> -required-
            revision    <string>
            singleBranch        <boolean>
            sshPrivateKeySecret <Object>
              key       <string> -required-
              name      <string>
              optional  <boolean>
            usernameSecret      <Object>
              key       <string> -required-
              name      <string>
              optional  <boolean>
          globalName    <string>
          hdfs  <Object>
            addresses   <[]string>
            dataTransferProtection      <string>
            force       <boolean>
            hdfsUser    <string>
            krbCCacheSecret     <Object>
              key       <string> -required-
              name      <string>
              optional  <boolean>
            krbConfigConfigMap  <Object>
              key       <string> -required-
              name      <string>
              optional  <boolean>
            krbKeytabSecret     <Object>
              key       <string> -required-
              name      <string>
              optional  <boolean>
            krbRealm    <string>
            krbServicePrincipalName     <string>
            krbUsername <string>
            path        <string> -required-
          http  <Object>
            auth        <Object>
              basicAuth <Object>
                passwordSecret  <Object>
                  key   <string> -required-
                  name  <string>
                  optional      <boolean>
                usernameSecret  <Object>
                  key   <string> -required-
                  name  <string>
                  optional      <boolean>
              clientCert        <Object>
                clientCertSecret        <Object>
                  key   <string> -required-
                  name  <string>
                  optional      <boolean>
                clientKeySecret <Object>
                  key   <string> -required-
                  name  <string>
                  optional      <boolean>
              oauth2    <Object>
                clientIDSecret  <Object>
                  key   <string> -required-
                  name  <string>
                  optional      <boolean>
                clientSecretSecret      <Object>
                  key   <string> -required-
                  name  <string>
                  optional      <boolean>
                endpointParams  <[]Object>
                  key   <string> -required-
                  value <string>
                scopes  <[]string>
                tokenURLSecret  <Object>
                  key   <string> -required-
                  name  <string>
                  optional      <boolean>
            headers     <[]Object>
              name      <string> -required-
              value     <string> -required-
            url <string> -required-
          mode  <integer>
          name  <string> -required-
          optional      <boolean>
          oss   <Object>
            accessKeySecret     <Object>
              key       <string> -required-
              name      <string>
              optional  <boolean>
            bucket      <string>
            createBucketIfNotPresent    <boolean>
            endpoint    <string>
            key <string> -required-
            lifecycleRule       <Object>
              markDeletionAfterDays     <integer>
              markInfrequentAccessAfterDays     <integer>
            secretKeySecret     <Object>
              key       <string> -required-
              name      <string>
              optional  <boolean>
            securityToken       <string>
            useSDKCreds <boolean>
          path  <string>
          raw   <Object>
            data        <string> -required-
          recurseMode   <boolean>
          s3    <Object>
            accessKeySecret     <Object>
              key       <string> -required-
              name      <string>
              optional  <boolean>
            bucket      <string>
            caSecret    <Object>
              key       <string> -required-
              name      <string>
              optional  <boolean>
            createBucketIfNotPresent    <Object>
              objectLocking     <boolean>
            encryptionOptions   <Object>
              enableEncryption  <boolean>
              kmsEncryptionContext      <string>
              kmsKeyId  <string>
              serverSideCustomerKeySecret       <Object>
                key     <string> -required-
                name    <string>
                optional        <boolean>
            endpoint    <string>
            insecure    <boolean>
            key <string>
            region      <string>
            roleARN     <string>
            secretKeySecret     <Object>
              key       <string> -required-
              name      <string>
              optional  <boolean>
            sessionTokenSecret  <Object>
              key       <string> -required-
              name      <string>
              optional  <boolean>
            useSDKCreds <boolean>
          subPath       <string>
        parameters      <[]Object>
          default       <string>
          description   <string>
          enum  <[]string>
          globalName    <string>
          name  <string> -required-
          value <string>
          valueFrom     <Object>
            configMapKeyRef     <Object>
              key       <string> -required-
              name      <string>
              optional  <boolean>
            default     <string>
            event       <string>
            expression  <string>
            jqFilter    <string>
            jsonPath    <string>
            parameter   <string>
            path        <string>
            supplied    <Object>
      metadata  <Object>
      workflowTemplateRef       <Object> -required-
        clusterScope    <boolean>
        name    <string>
```
</details>

<details>
<summary>Click here for <code>kubectl explain --recursive WorkflowArtifactGCTask</code> output</summary>

```
$ kubectl explain --recursive WorkflowArtifactGCTask
GROUP:      argoproj.io
KIND:       WorkflowArtifactGCTask
VERSION:    v1alpha1

DESCRIPTION:
    <empty>
FIELDS:
  apiVersion    <string>
  kind  <string>
  metadata      <ObjectMeta> -required-
    annotations <map[string]string>
    creationTimestamp   <string>
    deletionGracePeriodSeconds  <integer>
    deletionTimestamp   <string>
    finalizers  <[]string>
    generateName        <string>
    generation  <integer>
    labels      <map[string]string>
    managedFields       <[]ManagedFieldsEntry>
      apiVersion        <string>
      fieldsType        <string>
      fieldsV1  <FieldsV1>
      manager   <string>
      operation <string>
      subresource       <string>
      time      <string>
    name        <string>
    namespace   <string>
    ownerReferences     <[]OwnerReference>
      apiVersion        <string> -required-
      blockOwnerDeletion        <boolean>
      controller        <boolean>
      kind      <string> -required-
      name      <string> -required-
      uid       <string> -required-
    resourceVersion     <string>
    selfLink    <string>
    uid <string>
  spec  <Object> -required-
    artifactsByNode     <map[string]Object>
      archiveLocation   <Object>
        archiveLogs     <boolean>
        artifactory     <Object>
          passwordSecret        <Object>
            key <string> -required-
            name        <string>
            optional    <boolean>
          url   <string> -required-
          usernameSecret        <Object>
            key <string> -required-
            name        <string>
            optional    <boolean>
        azure   <Object>
          accountKeySecret      <Object>
            key <string> -required-
            name        <string>
            optional    <boolean>
          blob  <string> -required-
          container     <string> -required-
          endpoint      <string> -required-
          useSDKCreds   <boolean>
        gcs     <Object>
          bucket        <string>
          key   <string> -required-
          serviceAccountKeySecret       <Object>
            key <string> -required-
            name        <string>
            optional    <boolean>
        git     <Object>
          branch        <string>
          depth <integer>
          disableSubmodules     <boolean>
          fetch <[]string>
          insecureIgnoreHostKey <boolean>
          passwordSecret        <Object>
            key <string> -required-
            name        <string>
            optional    <boolean>
          repo  <string> -required-
          revision      <string>
          singleBranch  <boolean>
          sshPrivateKeySecret   <Object>
            key <string> -required-
            name        <string>
            optional    <boolean>
          usernameSecret        <Object>
            key <string> -required-
            name        <string>
            optional    <boolean>
        hdfs    <Object>
          addresses     <[]string>
          dataTransferProtection        <string>
          force <boolean>
          hdfsUser      <string>
          krbCCacheSecret       <Object>
            key <string> -required-
            name        <string>
            optional    <boolean>
          krbConfigConfigMap    <Object>
            key <string> -required-
            name        <string>
            optional    <boolean>
          krbKeytabSecret       <Object>
            key <string> -required-
            name        <string>
            optional    <boolean>
          krbRealm      <string>
          krbServicePrincipalName       <string>
          krbUsername   <string>
          path  <string> -required-
        http    <Object>
          auth  <Object>
            basicAuth   <Object>
              passwordSecret    <Object>
                key     <string> -required-
                name    <string>
                optional        <boolean>
              usernameSecret    <Object>
                key     <string> -required-
                name    <string>
                optional        <boolean>
            clientCert  <Object>
              clientCertSecret  <Object>
                key     <string> -required-
                name    <string>
                optional        <boolean>
              clientKeySecret   <Object>
                key     <string> -required-
                name    <string>
                optional        <boolean>
            oauth2      <Object>
              clientIDSecret    <Object>
                key     <string> -required-
                name    <string>
                optional        <boolean>
              clientSecretSecret        <Object>
                key     <string> -required-
                name    <string>
                optional        <boolean>
              endpointParams    <[]Object>
                key     <string> -required-
                value   <string>
              scopes    <[]string>
              tokenURLSecret    <Object>
                key     <string> -required-
                name    <string>
                optional        <boolean>
          headers       <[]Object>
            name        <string> -required-
            value       <string> -required-
          url   <string> -required-
        oss     <Object>
          accessKeySecret       <Object>
            key <string> -required-
            name        <string>
            optional    <boolean>
          bucket        <string>
          createBucketIfNotPresent      <boolean>
          endpoint      <string>
          key   <string> -required-
          lifecycleRule <Object>
            markDeletionAfterDays       <integer>
            markInfrequentAccessAfterDays       <integer>
          secretKeySecret       <Object>
            key <string> -required-
            name        <string>
            optional    <boolean>
          securityToken <string>
          useSDKCreds   <boolean>
        raw     <Object>
          data  <string> -required-
        s3      <Object>
          accessKeySecret       <Object>
            key <string> -required-
            name        <string>
            optional    <boolean>
          bucket        <string>
          caSecret      <Object>
            key <string> -required-
            name        <string>
            optional    <boolean>
          createBucketIfNotPresent      <Object>
            objectLocking       <boolean>
          encryptionOptions     <Object>
            enableEncryption    <boolean>
            kmsEncryptionContext        <string>
            kmsKeyId    <string>
            serverSideCustomerKeySecret <Object>
              key       <string> -required-
              name      <string>
              optional  <boolean>
          endpoint      <string>
          insecure      <boolean>
          key   <string>
          region        <string>
          roleARN       <string>
          secretKeySecret       <Object>
            key <string> -required-
            name        <string>
            optional    <boolean>
          sessionTokenSecret    <Object>
            key <string> -required-
            name        <string>
            optional    <boolean>
          useSDKCreds   <boolean>
      artifacts <map[string]Object>
        archive <Object>
          none  <Object>
          tar   <Object>
            compressionLevel    <integer>
          zip   <Object>
        archiveLogs     <boolean>
        artifactGC      <Object>
          podMetadata   <Object>
            annotations <map[string]string>
            labels      <map[string]string>
          serviceAccountName    <string>
          strategy      <string>
        artifactory     <Object>
          passwordSecret        <Object>
            key <string> -required-
            name        <string>
            optional    <boolean>
          url   <string> -required-
          usernameSecret        <Object>
            key <string> -required-
            name        <string>
            optional    <boolean>
        azure   <Object>
          accountKeySecret      <Object>
            key <string> -required-
            name        <string>
            optional    <boolean>
          blob  <string> -required-
          container     <string> -required-
          endpoint      <string> -required-
          useSDKCreds   <boolean>
        deleted <boolean>
        from    <string>
        fromExpression  <string>
        gcs     <Object>
          bucket        <string>
          key   <string> -required-
          serviceAccountKeySecret       <Object>
            key <string> -required-
            name        <string>
            optional    <boolean>
        git     <Object>
          branch        <string>
          depth <integer>
          disableSubmodules     <boolean>
          fetch <[]string>
          insecureIgnoreHostKey <boolean>
          passwordSecret        <Object>
            key <string> -required-
            name        <string>
            optional    <boolean>
          repo  <string> -required-
          revision      <string>
          singleBranch  <boolean>
          sshPrivateKeySecret   <Object>
            key <string> -required-
            name        <string>
            optional    <boolean>
          usernameSecret        <Object>
            key <string> -required-
            name        <string>
            optional    <boolean>
        globalName      <string>
        hdfs    <Object>
          addresses     <[]string>
          dataTransferProtection        <string>
          force <boolean>
          hdfsUser      <string>
          krbCCacheSecret       <Object>
            key <string> -required-
            name        <string>
            optional    <boolean>
          krbConfigConfigMap    <Object>
            key <string> -required-
            name        <string>
            optional    <boolean>
          krbKeytabSecret       <Object>
            key <string> -required-
            name        <string>
            optional    <boolean>
          krbRealm      <string>
          krbServicePrincipalName       <string>
          krbUsername   <string>
          path  <string> -required-
        http    <Object>
          auth  <Object>
            basicAuth   <Object>
              passwordSecret    <Object>
                key     <string> -required-
                name    <string>
                optional        <boolean>
              usernameSecret    <Object>
                key     <string> -required-
                name    <string>
                optional        <boolean>
            clientCert  <Object>
              clientCertSecret  <Object>
                key     <string> -required-
                name    <string>
                optional        <boolean>
              clientKeySecret   <Object>
                key     <string> -required-
                name    <string>
                optional        <boolean>
            oauth2      <Object>
              clientIDSecret    <Object>
                key     <string> -required-
                name    <string>
                optional        <boolean>
              clientSecretSecret        <Object>
                key     <string> -required-
                name    <string>
                optional        <boolean>
              endpointParams    <[]Object>
                key     <string> -required-
                value   <string>
              scopes    <[]string>
              tokenURLSecret    <Object>
                key     <string> -required-
                name    <string>
                optional        <boolean>
          headers       <[]Object>
            name        <string> -required-
            value       <string> -required-
          url   <string> -required-
        mode    <integer>
        name    <string> -required-
        optional        <boolean>
        oss     <Object>
          accessKeySecret       <Object>
            key <string> -required-
            name        <string>
            optional    <boolean>
          bucket        <string>
          createBucketIfNotPresent      <boolean>
          endpoint      <string>
          key   <string> -required-
          lifecycleRule <Object>
            markDeletionAfterDays       <integer>
            markInfrequentAccessAfterDays       <integer>
          secretKeySecret       <Object>
            key <string> -required-
            name        <string>
            optional    <boolean>
          securityToken <string>
          useSDKCreds   <boolean>
        path    <string>
        raw     <Object>
          data  <string> -required-
        recurseMode     <boolean>
        s3      <Object>
          accessKeySecret       <Object>
            key <string> -required-
            name        <string>
            optional    <boolean>
          bucket        <string>
          caSecret      <Object>
            key <string> -required-
            name        <string>
            optional    <boolean>
          createBucketIfNotPresent      <Object>
            objectLocking       <boolean>
          encryptionOptions     <Object>
            enableEncryption    <boolean>
            kmsEncryptionContext        <string>
            kmsKeyId    <string>
            serverSideCustomerKeySecret <Object>
              key       <string> -required-
              name      <string>
              optional  <boolean>
          endpoint      <string>
          insecure      <boolean>
          key   <string>
          region        <string>
          roleARN       <string>
          secretKeySecret       <Object>
            key <string> -required-
            name        <string>
            optional    <boolean>
          sessionTokenSecret    <Object>
            key <string> -required-
            name        <string>
            optional    <boolean>
          useSDKCreds   <boolean>
        subPath <string>
  status        <Object>
    artifactResultsByNode       <map[string]Object>
      artifactResults   <map[string]Object>
        error   <string>
        name    <string> -required-
        success <boolean>
```
</details>

### CODE REVIEW

**Review by agilgur5 — COMMENTED** [review]
> However, `WorkflowEventBinding` and `WorkflowArtifactGCTask` are small enough that minimization isn't necessary.

For now. I think it would be good to add a size check in the script so there are no surprises in the future

File: manifests/base/crds/full/README.md [hunk 1]
```diff
@@ -1,3 +1,3 @@
 # Full CRDs
 
-These CRDs have full schema validation. As a result, they are large and probably not suitable to be used in your cluster.
+These CRDs have the full schema. As a result, they are large and probably not suitable to be used in your cluster.
```
[the remark this answers is not in the document]
Comment (MasonM) [reply manifests/base/crds/full/README.md]: Okay, done in a2ca08019bb42ce5301677de61edc199968589ea

File: manifests/base/crds/minimal/README.md [hunk 2]
```diff
@@ -1,3 +1,3 @@
 # Minimal CRDs
 
-These CRDs omit schema validation.
+These CRDs omit large fields as a workaround for [Large CRDs go over size limits (e.g. those with embedded podspecs)](https://github.com/kubernetes/kubernetes/issues/82292).
```
[the remark this answers is not in the document]
Comment (MasonM) [reply manifests/base/crds/minimal/README.md]: Okay, done in a2ca08019bb42ce5301677de61edc199968589ea

**Comment by MasonM** [comment]
@agilgur5 Thanks for the review! I added additional safeguards in a2ca08019bb42ce5301677de61edc199968589ea. I also discovered there's a secondary issue preventing us from generating the full CRDs for `workflows.argoproj.io` and a few others: k8s rejects it with errors like this:
```
The CustomResourceDefinition "workflowtasksets.argoproj.io" is invalid: 
* spec.validation.openAPIV3Schema.properties[spec].properties[tasks].additionalProperties.properties[dag].properties[tasks].items.properties[inline].type: Required value: must not be empty for specified object fields
* spec.validation.openAPIV3Schema.properties[spec].properties[tasks].additionalProperties.properties[steps].items.items: Required value: must be specifiedk
```

I think this is related to the kubebuilder issues mentioned at https://github.com/argoproj/argo-workflows/pull/3809#discussion_r472383090

**Review by agilgur5 — APPROVED** [review]
LGTM. I like the two safeguards too

**Pull request merged**